konfetti/src/main/java/nl/dionsegijn/konfetti/models/Shape.kt

```diff
@@ -1,12 +1,93 @@
 package nl.dionsegijn.konfetti.models
 
-/**
- * Created by dionsegijn on 3/26/17.
- * Available shapes
- * RECT (rectangle)
- * Circle
- */
-enum class Shape {
-    RECT(),
-    CIRCLE()
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.PorterDuff
+import android.graphics.RectF
+import android.graphics.drawable.Drawable
+
+interface Shape {
+    /**
+     * Draw a shape to `canvas`. Implementations are expected to draw within a square of size
+     * `size` and must vertically/horizontally center their asset if it does not have an equal width
+     * and height.
+     */
+    fun draw(canvas: Canvas, paint: Paint, size: Float)
+
+    companion object {
+        // Maintain binary and backwards compatibility with previous enum API.
+        @JvmField
+        @Deprecated("Use Square class, instead.", replaceWith = ReplaceWith("Shape.Square"))
+        val RECT = Square
+
+        @JvmField
+        @Deprecated("Use Circle class, instead.", replaceWith = ReplaceWith("Shape.Circle"))
+        val CIRCLE = Circle
+    }
+
+    object Square : Shape {
+        override fun draw(
+            canvas: Canvas,
+            paint: Paint,
+            size: Float
+        ) {
+            canvas.drawRect(0f, 0f, size, size, paint)
+        }
+    }
+
+    class Rectangle(
+        /** The ratio of height to width. Must be within range [0, 1] */
+        private val sizeRatio: Float
+    ) : Shape {
+        init {
+            require(sizeRatio in 0f..1f)
+        }
+
+        override fun draw(canvas: Canvas, paint: Paint, size: Float) {
+            val height = size * sizeRatio
+            val top = (size - height) / 2f
+            canvas.drawRect(0f, top, size, top + height, paint)
+        }
+    }
+
+    object Circle : Shape {
+        private val rect = RectF()
+
+        override fun draw(
+            canvas: Canvas,
+            paint: Paint,
+            size: Float
+        ) {
+            rect.set(0f, 0f, size, size)
+            canvas.drawOval(rect, paint)
+        }
+    }
+
+    class DrawableShape(
+        drawable: Drawable,
+        /** Set to `false` to opt out of tinting. */
+        private val tint: Boolean = true
+    ) : Shape {
+        private val drawable = drawable.mutate()
+        private val sizeRatio =
+            if (drawable.intrinsicHeight == -1 || drawable.intrinsicWidth == -1) {
```

Comment:
Worth pointing out that I don't know when this would happen and am not sure if this is the right value to use for that case.

Reply:
How about using `getMinimumWidth`? The fallback in this function is 0. 
This will result in a drawable with a height of 0 later in the drawing operation. Assuming an intrinsicHeight of -1 is not good it won't draw a visible image here. 

Another option is to warn the developer here that the drawable used is probably not correct. Though I'd rather draw nothing than a runtime crash if it's caused by something else and ends up in a production app.

![image](https://user-images.githubusercontent.com/1636897/76111242-384b6680-5fe0-11ea-8ae5-1e7cf3e4dd2e.png)

Reply:
I updated this logic. If they're both `-1` then I return `1f` to fill the space, but if only one of them is `-1` then I return `0f` to prevent drawing. I considered a change where we can handle only one being `-1`, but I think it was too complicated without a real use case to evaluate it on. Posting here in case you want to consider it in the future:

```kt
private fun calculateHeightForSize(size: Float): Float {
    return if (drawable.intrinsicHeight == -1 && drawable.intrinsicWidth == -1) {
        // If the drawable has no intrinsic size, fill the available space.
        size
    } else if (drawable.intrinsicWidth == -1) {
        // The drawable has no intrinsic width, so it is meant to have a fixed height
        // regardless of the width. Return the intrinsic height.
        drawable.intrinsicHeight.toFloat()
    } else if (drawable.intrinsicHeight == -1) {
        // The drawable has no intrinsic height, so it is meant to have a fixed width. We
        // don't support this right now, so return 0.
        0f
    } else {
        size * drawable.intrinsicHeight.toFloat() / drawable.intrinsicWidth
    }
}
```